### PR TITLE
feat: added external 3ds support using netcetera

### PR DESCRIPTION
**Pull request by Sanskar2001**
Added Redirection less 3DS Auth support using Netcetera

### CODE REVIEW

File: src/hooks/LoggerHook.res [hunk 1]
```diff
@@ -32,6 +33,14 @@ type eventName =
   | PAYMENT_SESSION_INITIATED
   | LOADER_CHANGED
   | SCAN_CARD
+  | AUTHENTICATION_CALL_INIT
+  | AUTHENTICATION_CALL
+  | AUTHORISE_CALL_INIT
```
Comment (vsrivatsa-juspay) [inline src/hooks/LoggerHook.res:38]: AUTHORIZE (coz American lmao)

File: src/hooks/LoggerHook.res [hunk 2]
```diff
@@ -86,6 +95,14 @@ let eventToStrMapper = eventName => {
   | PAYMENT_SESSION_INITIATED => "PAYMENT_SESSION_INITIATED"
   | LOADER_CHANGED => "LOADER_CHANGED"
   | SCAN_CARD => "SCAN_CARD"
+  | AUTHENTICATION_CALL_INIT => " AUTHENTICATION_CALL_INIT"
+  | AUTHENTICATION_CALL => " AUTHENTICATION_CALL"
+  | AUTHORIZE_CALL_INIT => " AUTHORIZE_CALL_INIT"
```
Comment (vsrivatsa-juspay) [inline src/hooks/LoggerHook.res:100]:
```suggestion
  | AUTHORIZE_CALL_INIT => "AUTHORIZE_CALL_INIT"
```

File: src/hooks/LoggerHook.res [hunk 3]
```diff
@@ -86,6 +95,14 @@ let eventToStrMapper = eventName => {
...
+  | AUTHORIZE_CALL => " AUTHORIZE_CALL"
```
Comment (vsrivatsa-juspay) [inline src/hooks/LoggerHook.res:101]:
```suggestion
  | AUTHORIZE_CALL => "AUTHORIZE_CALL"
```

File: src/hooks/LoggerHook.res [hunk 4]
```diff
@@ -86,6 +95,14 @@ let eventToStrMapper = eventName => {
...
+  | POLL_STATUS_CALL_INIT => " POLL_STATUS_CALL_INIT"
+  | POLL_STATUS_CALL => " POLL_STATUS_CALL"
```
Comment (vsrivatsa-juspay) [inline src/hooks/LoggerHook.res:103]:
```suggestion
  | POLL_STATUS_CALL => "POLL_STATUS_CALL"
```

File: src/hooks/LoggerHook.res [hunk 5]
```diff
@@ -86,6 +95,14 @@ let eventToStrMapper = eventName => {
...
+  | DISPLAY_THREE_DS_SDK => " DISPLAY_THREE_DS_SDK"
```
Comment (vsrivatsa-juspay) [inline src/hooks/LoggerHook.res:104]:
```suggestion
  | DISPLAY_THREE_DS_SDK => "DISPLAY_THREE_DS_SDK"
```

File: src/hooks/LoggerHook.res [hunk 6]
```diff
@@ -86,6 +95,14 @@ let eventToStrMapper = eventName => {
...
+  | NETCETERA_SDK => " NETCETERA_SDK"
```
Comment (vsrivatsa-juspay) [inline src/hooks/LoggerHook.res:105]:
```suggestion
  | NETCETERA_SDK => "NETCETERA_SDK"
```

**Review by vsrivatsa-juspay — CHANGES_REQUESTED** [review]
Whitespaces to be removed in event names

File: src/hooks/NetceteraThreeDsHooks.res [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+        ~logType=INFO,
+        ~eventName=RETRIEVE_CALL_INIT,
+        ~url=baseUrl,
+        ~statusCode="",
+        ~apiLogType=Request,
+        ~data=JSON.Encode.null,
+        (),
+      )
+      retrievePayment(Types.Payment, clientSecret, publishableKey)
+      ->Promise.then(res => {
```
Comment (vsrivatsa-juspay) [inline src/hooks/NetceteraThreeDsHooks.res:30]: add catch block for the promise

File: src/hooks/NetceteraThreeDsHooks.res [hunk 8]
```diff
@@ -0,0 +1,443 @@
...
+        if pollCounter.contents >= pollConfig.frequency {
+          clearInterval(intervalId.contents)
+          if paymentStatus.contents !== "completed" {
+            onSuccess()
+          }
+        }
+
+        let headers = getAuthCallHeaders(publishableKey)
+        CommonHooks.fetchApi(~uri, ~headers, ~method_=Fetch.Get, ())
+        ->Promise.then(data => {
```
Comment (vsrivatsa-juspay) [inline src/hooks/NetceteraThreeDsHooks.res:88]: add catch block for the promise

File: src/hooks/NetceteraThreeDsHooks.res [hunk 9]
```diff
@@ -0,0 +1,443 @@
...
+        (),
+      )
+      let headers = [("Content-Type", "application/json")]->Dict.fromArray
+      CommonHooks.fetchApi(
+        ~uri=authoriseUrl,
+        ~bodyStr="",
+        ~headers,
+        ~method_=Fetch.Post,
+        (),
+      )->Promise.then(data => {
```
Comment (vsrivatsa-juspay) [inline src/hooks/NetceteraThreeDsHooks.res:165]: add catch block for the promise

File: src/hooks/NetceteraThreeDsHooks.res [hunk 10]
```diff
@@ -0,0 +1,443 @@
...
+        ~eventName=AUTHENTICATION_CALL_INIT,
+        ~url=uri,
+        ~statusCode="",
+        ~apiLogType=Request,
+        ~data=JSON.Encode.null,
+        (),
+      )
+
+      CommonHooks.fetchApi(~uri, ~bodyStr, ~headers, ~method_=Post, ())
+      ->Promise.then(data => {
```
Comment (vsrivatsa-juspay) [inline src/hooks/NetceteraThreeDsHooks.res:282]: add catch block for the promise

File: src/hooks/NetceteraThreeDsHooks.res [hunk 11]
```diff
@@ -0,0 +1,443 @@
...
+              ~statusCode,
+              ~apiLogType=Response,
+              ~data=JSON.Encode.null,
+              (),
+            )
+            let authResponse = res->authResponseItemToObjMapper
+
+            switch authResponse {
+            | AUTH_RESPONSE(challengeParams) =>
+              logger(
```
Comment (vsrivatsa-juspay) [inline src/hooks/NetceteraThreeDsHooks.res:301]: move this log outside the switch case. We would want to see the value even if its not "C"

File: src/types/ExternalThreeDsTypes.res [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    ? AUTH_ERROR({
+        errorCode: dict->ErrorUtils.getErrorCode,
+        errorMessage: dict->ErrorUtils.getErrorMessage,
+      })
+    : AUTH_RESPONSE({
+        transStatus: dict->Utils.getDictFromJson->Utils.getString("trans_status", ""),
+        acsSignedContent: dict->Utils.getDictFromJson->Utils.getString("acs_signed_content", ""),
+        acsRefNumber: dict->Utils.getDictFromJson->Utils.getString("acs_reference_number", ""),
+        acsTransactionId: dict->Utils.getDictFromJson->Utils.getString("acs_trans_id", ""),
+        threeDSRequestorURL: "https://example.com/3ds-requestor-url",
```
Comment (vsrivatsa-juspay) [inline src/types/ExternalThreeDsTypes.res:69]: why is this hardcoded?

File: src/utility/constants/SdkStatusMessages.res [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+  apiCallFailure: "polling failure, cannot fetch the status of payment",
+}
+
+let externalThreeDsModuleStatus = {
+  successMsg: "external 3DS dependency found",
+  errorMsg: "integration error, external 3DS dependency not found",
+  apiCallFailure: "",
+}
+
+let authoriseCallStatus = {
```
Comment (vsrivatsa-juspay) [inline src/utility/constants/SdkStatusMessages.res:25]:
```suggestion
let authorizeCallStatus = {
```

File: src/utility/constants/SdkStatusMessages.res [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+  successMsg: "payment authorised successfully",
+  errorMsg: "authorise failed",
```
Comment (vsrivatsa-juspay) [inline src/utility/constants/SdkStatusMessages.res:27]:
```suggestion
  errorMsg: "authorize failed",
```

File: src/utility/constants/SdkStatusMessages.res [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+  apiCallFailure: "authorise failure, cannot process this payment",
```
Comment (vsrivatsa-juspay) [inline src/utility/constants/SdkStatusMessages.res:28]:
```suggestion
  apiCallFailure: "authorize failure, cannot process this payment",
```

**Review by vsrivatsa-juspay — DISMISSED** [review]

File: src/hooks/NetceteraThreeDsHooks.res [hunk 16]
```diff
@@ -0,0 +1,476 @@
...
+          threeDsData.directoryServerId,
+          status => {
+            logger(
+              ~logType=INFO,
+              ~value=status->JSON.stringifyAny->Option.getOr(""),
+              ~category=USER_EVENT,
+              ~eventName=NETCETERA_SDK,
+              (),
+            )
+            status->isStatusSuccess
```
Comment (vsrivatsa-juspay) [inline src/hooks/NetceteraThreeDsHooks.res:398]: if else condition

File: src/hooks/NetceteraThreeDsHooks.res [hunk 17]
```diff
@@ -0,0 +1,476 @@
...
+
+      let onChallengeCompletionCallback = message => {
+        retrievePayment(Payment, clientSecret, publishableKey)
+        ->Promise.then(res => {
+          if res == JSON.Encode.null {
+            onFailure(retrievePaymentStatus.apiCallFailure)
+          } else {
+            let status = res->Utils.getDictFromJson->Utils.getString("status", "")
+            switch status {
+            | "processing" => onSuccess(retrievePaymentStatus.successMsg)
```
Comment (vsrivatsa-juspay) [inline src/hooks/NetceteraThreeDsHooks.res:449]:
```suggestion
            | "processing"
```

File: src/utility/constants/SdkStatusMessages.res [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+let externalThreeDsModuleStatus = {
+  successMsg: "external 3DS dependency found",
+  errorMsg: "integration error, external 3DS dependency not found",
+  apiCallFailure: "",
+}
+
+let authorizeCallStatus = {
+  successMsg: "payment authorised successfully",
```
Comment (vsrivatsa-juspay) [inline src/utility/constants/SdkStatusMessages.res:26]:
```suggestion
  successMsg: "payment authorized successfully",
```

**Review by aashu331998 — APPROVED** [review]

**Pull request merged**